File: user-service/src/test/java/com/example/user/service/UserServiceTest.java
package com.example.user.service;

import com.example.user.model.User;
import com.example.user.repository.UserRepository;
import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.Test;
import org.mockito.InjectMocks;
import org.mockito.Mock;
import org.mockito.MockitoAnnotations;

import java.util.Optional;

import static org.junit.jupiter.api.Assertions.*;
import static org.mockito.ArgumentMatchers.any;
import static org.mockito.Mockito.*;

class UserServiceTest {
    
    @Mock
    private UserRepository userRepository;
    
    @InjectMocks
    private UserService userService;
    
    @BeforeEach
    void setUp() {
        MockitoAnnotations.openMocks(this);
    }
    
    @Test
    void shouldCreateUserWhenEmailAndUsernameAreUnique() {
        // Given
        User user = new User("testuser", "test@example.com", "password");
        User savedUser = new User("testuser", "test@example.com", "password");
        savedUser.setId(1L);
        
        when(userRepository.findByEmail("test@example.com")).thenReturn(Optional.empty());
        when(userRepository.findByUsername("testuser")).thenReturn(Optional.empty());
        when(userRepository.save(any(User.class))).thenReturn(savedUser);
        
        // When
        User result = userService.createUser(user);
        
        // Then
        assertNotNull(result);
        assertEquals(1L, result.getId());
        assertEquals("testuser", result.getUsername());
        assertEquals("test@example.com", result.getEmail());
        
        verify(userRepository).findByEmail("test@example.com");
        verify(userRepository).findByUsername("testuser");
        verify(userRepository).save(any(User.class));
    }
    
    @Test
    void shouldThrowExceptionWhenEmailAlreadyExists() {
        // Given
        User user = new User("testuser", "test@example.com", "password");
        User existingUser = new User("otheruser", "test@example.com", "password");
        existingUser.setId(1L);
        
        when(userRepository.findByEmail("test@example.com")).thenReturn(Optional.of(existingUser));
        
        // When & Then
        RuntimeException exception = assertThrows(RuntimeException.class, () -> {
            userService.createUser(user);
        });
        
        assertEquals("User with email test@example.com already exists", exception.getMessage());
        verify(userRepository).findByEmail("test@example.com");
        verify(userRepository, never()).findByUsername(anyString());
        verify(userRepository, never()).save(any(User.class));
    }
    
    @Test
    void shouldGetUserById() {
        // Given
        Long id = 1L;
        User user = new User("testuser", "test@example.com", "password");
        user.setId(id);
        
        when(userRepository.findById(id)).thenReturn(Optional.of(user));
        
        // When
        Optional<User> result = userService.getUserById(id);
        
        // Then
        assertTrue(result.isPresent());
        assertEquals(id, result.get().getId());
        assertEquals("testuser", result.get().getUsername());
        
        verify(userRepository).findById(id);
    }
}

File: user-service/src/test/kotlin/com/example/user/service/UserServiceTest.kt
package com.example.user.service

import com.example.user.model.User
import com.example.user.repository.UserRepository
import io.mockk.*
import io.mockk.impl.annotations.MockK
import org.junit.jupiter.api.BeforeEach
import org.junit.jupiter.api.Test
import org.junit.jupiter.api.Assertions.*
import org.springframework.test.util.ReflectionTestUtils
import java.util.*

class UserServiceTest {
    
    @MockK
    private lateinit var userRepository: UserRepository
    
    private lateinit var userService: UserService
    
    @BeforeEach
    fun setUp() {
        MockKAnnotations.init(this)
        userService = UserService(userRepository)
    }
    
    @Test
    fun `should create user when email and username are unique`() {
        // Given
        val user = User(username = "testuser", email = "test@example.com", password = "password")
        val savedUser = User(id = 1, username = "testuser", email = "test@example.com", password = "password")
        every { userRepository.findByEmail(any()) } returns Optional.empty()
        every { userRepository.findByUsername(any()) } returns Optional.empty()
        every { userRepository.save(any()) } returns savedUser
        
        // When
        val result = userService.createUser(user)
        
        // Then
        assertEquals(savedUser, result)
        verify { userRepository.findByEmail("test@example.com") }
        verify { userRepository.findByUsername("testuser") }
        verify { userRepository.save(match { it.username == "testuser" && it.email == "test@example.com" }) }
    }
    
    @Test
    fun `should throw exception when email already exists`() {
        // Given
        val user = User(username = "testuser", email = "test@example.com", password = "password")
        val existingUser = User(id = 1, username = "otheruser", email = "test@example.com", password = "password")
        every { userRepository.findByEmail("test@example.com") } returns Optional.of(existingUser)
        
        // When/Then
        assertThrows(RuntimeException::class.java) {
            userService.createUser(user)
        }
        verify { userRepository.findByEmail("test@example.com") }
        verify(exactly = 0) { userRepository.findByUsername(any()) }
    }
    
    @Test
    fun `should throw exception when username already exists`() {
        // Given
        val user = User(username = "testuser", email = "test@example.com", password = "password")
        val existingUser = User(id = 1, username = "testuser", email = "other@example.com", password = "password")
        every { userRepository.findByEmail(any()) } returns Optional.empty()
        every { userRepository.findByUsername("testuser") } returns Optional.of(existingUser)
        
        // When/Then
        assertThrows(RuntimeException::class.java) {
            userService.createUser(user)
        }
        verify { userRepository.findByEmail("test@example.com") }
        verify { userRepository.findByUsername("testuser") }
    }
    
    @Test
    fun `should find user by id`() {
        // Given
        val id = 1L
        val user = User(id = id, username = "testuser", email = "test@example.com", password = "password")
        every { userRepository.findById(id) } returns Optional.of(user)
        
        // When
        val result = userService.getUserById(id)
        
        // Then
        assertTrue(result.isPresent)
        assertEquals(user, result.get())
        verify { userRepository.findById(id) }
    }
    
    @Test
    fun `should return empty optional when user not found by id`() {
        // Given
        val id = 1L
        every { userRepository.findById(id) } returns Optional.empty()
        
        // When
        val result = userService.getUserById(id)
        
        // Then
        assertFalse(result.isPresent)
        verify { userRepository.findById(id) }
    }
    
    @Test
    fun `should delete user successfully`() {
        // Given
        val id = 1L
        every { userRepository.existsById(id) } returns true
        every { userRepository.deleteById(id) } just Runs
        
        // When
        val result = userService.deleteUser(id)
        
        // Then
        assertTrue(result)
        verify { userRepository.existsById(id) }
        verify { userRepository.deleteById(id) }
    }
    
    @Test
    fun `should not delete user if not exists`() {
        // Given
        val id = 1L
        every { userRepository.existsById(id) } returns false
        
        // When
        val result = userService.deleteUser(id)
        
        // Then
        assertFalse(result)
        verify { userRepository.existsById(id) }
        verify(exactly = 0) { userRepository.deleteById(any()) }
    }
}